crates/pipeline_manager/src/compiler.rs

```diff
@@ -456,13 +472,101 @@ impl Compiler {
         })
     }
 
+    async fn reconcile_local_state(
+        config: &CompilerConfig,
+        db: &Arc<Mutex<ProjectDB>>,
+    ) -> Result<(), DBError> {
+        info!("Reconciling local state with API state");
+        let mut map: HashSet<(Uuid, i64)> = HashSet::new();
+        let read_dir = fs::read_dir(config.binaries_dir()).await;
+        match read_dir {
+            Ok(mut paths) => loop {
+                let entry = paths.next_entry().await;
+                match entry {
+                    Ok(Some(path)) => {
+                        let maybe_parts = Self::binary_path_to_parts(&path).await;
+                        match maybe_parts {
+                            Some((program_uuid, program_version)) => {
+                                map.insert((program_uuid.0, program_version.0));
+                            }
+                            None => {
+                                warn!(
+                                    "Local state reconciler found invalid file in {:?}: {:?}",
+                                    config.binaries_dir(),
+                                    path.file_name()
+                                );
+                            }
+                        }
+                    }
+                    // We are done with the directory
+                    Ok(None) => break,
+                    Err(e) => {
+                        warn!("Local state reconciler was unable to read an entry: {}", e);
+                        // Not clear from docs if an error during iteration
+                        // is recoverable and we could just `continue;` here
+                        break;
+                    }
+                }
+            },
+            Err(e) => {
+                warn!(
+                    "Local state reconciler could not read binaries directory: {}",
+                    e
+                )
+            }
+        }
+
+        let programs = db.lock().await.all_programs().await?;
+        for (tenant_id, program) in programs {
+            // We have an artifact but the program status has not been updated to Success. This
+            // could happen because the compiler service could not successfully reach the database
+            // after finishing the compilation job. TODO: it could also be that some other compiler
+            // instance is working on the same program, but we don't handle that deployment mode yet.
+            if program.status == ProgramStatus::CompilingRust
+                && map.contains(&(program.program_id.0, program.version.0))
+            {
+                db.lock()
+                    .await
+                    .set_program_status_guarded(
+                        tenant_id,
+                        program.program_id,
+                        program.version,
+                        ProgramStatus::Success,
```

Comment:
This seems risky.  The file may be partially written for example. If this happens, I'd rather re-compile the program.